Incorporate changes to workshop materials from old repo (#493)
DKISTDC/dkist: docs/tutorial/2_search_and_asdf_download.md

(dkist:tutorial:search-and-download)=
# Searching for DKIST Datasets

> In this session we will cover how to search for DKIST datasets available at the DKIST Data Center.

Reviewer: ```suggestion
In this chapter we will cover how to search for DKIST datasets available at the DKIST Data Center.
```

In DKIST data parlance, a "dataset" is the smallest unit of data that is searchable from the data centre, and represents a single observation from a single instrument at a single pass band.

Each dataset comprises a number of different files:
  * An ASDF file containing all the metadata, and no data.
  * A quality report PDF.
  * An mp4 preview movie.
  * A (large) number of FITS files, each containing a "calibrated exposure".

> All of these files apart from the FITS files containing the data can be downloaded irrespective of embargo status.

Reviewer: ```suggestion
All of these files, apart from the FITS files containing the data, can be downloaded irrespective of embargo status.
```

For each of these "datasets" the DKIST Data Center keeps a "dataset inventory record" which is a limited set of metadata about the dataset on which you can search, either through the web portal or the `dkist` Python package.
The ASDF, quality report and preview movie can all be downloaded without authenticating, the FITS files require the use of Globus.

## Using `Fido.search`

The search interface we are going to use is {obj}`sunpy.net.Fido`.
`Fido` supports many different sources of data, some built into `sunpy` like the VSO and some in plugins like `dkist` or `sunpy-soar`.
With `Fido` you can search for DKIST datasets and download their corresponding ASDF files.
To register the DKIST search with `Fido` we must also import `dkist.net`.

```{code-cell} python
import astropy.units as u
from sunpy.net import Fido, attrs as a
import dkist.net
```

> `Fido` searches are built up from "attrs", which we imported above as `a`.

Reviewer: ```suggestion
`Fido` searches are built up from attributes, referred to in python as `attrs`, which we have imported above as `a`.
```

If we only want VBI datasets, that are unembargoed, between a specific time range we can use multiple attrs:

```{code-cell} python
Fido.search(a.Time("2022-06-02 17:00:00", "2022-06-02 18:00:00") & a.Instrument.vbi & a.dkist.Embargoed(False))
```

Note how the `a.Time` and `a.Instrument` attrs are not prefixed with `dkist` - these are general attrs which can be used to search multiple sources of data.

So far all returned results have had to match all the attrs provided, because we have used the `&` (logical AND) operator to join them.
If we want results that match either one of multiple options we can use the `|` operator.

```{code-cell} python
---
tags: [output_scroll]
---
Fido.search((a.Instrument.vbi | a.Instrument.visp) & a.dkist.Embargoed(False))
```

As you can see this has returned two separate tables, one for VBI and one for VISP.

> Because `Fido` can search other clients as well as the DKIST you can make a more complex query which will search for VISP data and context images from AIA at the same time:

Reviewer: ```suggestion
Because `Fido` can search other clients in addition to the DKIST specific client. You can create more complex queries which will search for VISP data and context images from AIA at the same time:
```
